File: NimdaConBackEnd/backend-spring/src/main/java/com/nimda/con/service/AuthService.java
package com.nimda.con.service;

import com.nimda.con.entity.User;
import com.nimda.con.util.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public class AuthService {
    
    @Autowired
    private UserService userService;
    
    @Autowired
    private PasswordEncoder passwordEncoder;
    
    @Autowired
    private JwtUtil jwtUtil;
    
    /**
     * 사용자 인증
     * @param username 사용자명
     * @param password 비밀번호
     * @return 인증된 사용자 정보 (비밀번호 제외)
     */
    public Optional<User> validateUser(String username, String password) {
        Optional<User> userOpt = userService.findByUsername(username);
        
        if (userOpt.isPresent()) {
            User user = userOpt.get();
            if (passwordEncoder.matches(password, user.getPassword())) {
                // 비밀번호를 제외한 사용자 정보 반환
                User userWithoutPassword = new User();
                userWithoutPassword.setId(user.getId());
                userWithoutPassword.setUsername(user.getUsername());
                userWithoutPassword.setEmail(user.getEmail());
                userWithoutPassword.setCreatedAt(user.getCreatedAt());
                userWithoutPassword.setUpdatedAt(user.getUpdatedAt());
                return Optional.of(userWithoutPassword);
            }
        }
        
        return Optional.empty();
    }
    
    /**
     * 로그인 처리
     * @param user 인증된 사용자 정보
     * @return JWT 토큰이 포함된 응답
     */
    public Map<String, Object> login(User user) {
        Map<String, Object> response = new HashMap<>();
        
        // JWT 토큰 생성
        String token = jwtUtil.generateToken(user.getUsername(), user.getId());
        
        response.put("access_token", token);
        response.put("user", user);
        
        return response;
    }
    
    /**
     * 회원가입 처리
     * @param username 사용자명
     * @param password 비밀번호
     * @param email 이메일
     * @return 생성된 사용자 정보 (비밀번호 제외)
     */
    public User register(String username, String password, String email) {
        // 중복 사용자명 확인
        if (userService.existsByUsername(username)) {
            throw new RuntimeException("Username already exists");
        }
        
        // 중복 이메일 확인
        if (userService.existsByEmail(email)) {
            throw new RuntimeException("Email already exists");
        }
        
        // 새 사용자 생성
        User user = userService.createUser(username, password, email);
        
        // 비밀번호를 제외한 사용자 정보 반환
        User userWithoutPassword = new User();
        userWithoutPassword.setId(user.getId());
        userWithoutPassword.setUsername(user.getUsername());
        userWithoutPassword.setEmail(user.getEmail());
        userWithoutPassword.setCreatedAt(user.getCreatedAt());
        userWithoutPassword.setUpdatedAt(user.getUpdatedAt());
        
        return userWithoutPassword;
    }
}

File: NimdaConBackEnd/backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from '../users/users.service';
import * as bcrypt from 'bcryptjs';

@Injectable()
export class AuthService {

  // UsersService : 사용자 관리 도구 (관리자 계정 생성, 임시(메모리) DB 연동, 사용자 탐색)
  // JwtService : JWT 토큰 생성 도구 
  constructor(
    private usersService: UsersService,
    private jwtService: JwtService,
  ) {}

  // validateUser : 사용자 인증 로직 
  // 사용자명과 비밀번호를 받아서 사용자 정보를 탐색 
  // 사용자명이 일치하는 사용자 정보를 반환 (User or undefined)
  async validateUser(username: string, password: string): Promise<any> {

    // #1 사용자 이름으로 사용자 탐색 
    const user = await this.usersService.findByUsername(username);

    // #2 사용자 명과 해시화된 비밀번호를 비교 
    if (user && await bcrypt.compare(password, user.password)) {
      const { password, ...result } = user;
      return result;
    }

    // #3 검증 실패 
    return null;
  }

  // login : 사용자 로그인 처리 
  // 사용자 정보를 받아서 JWT 토큰 생성 
  // 생성된 토큰을 반환 
  async login(user: any) {
    const payload = { username: user.username, sub: user.id };
    return {
      access_token: this.jwtService.sign(payload),
    };
  }

  // register : 사용자 회원가입 처리
  // 사용자명, 비밀번호, 이메일을 받아서 새 사용자 생성
  // 생성된 사용자 정보를 반환

  async register(username: string, password: string, email: string): Promise<any> {
    
    // #1 중복 사용자명 확인
    const existingUser = await this.usersService.findByUsername(username);
    if (existingUser) {
      throw new UnauthorizedException('Username already exists');
    }

    // #2 새 사용자 생성
    const user = await this.usersService.create(username, password, email);
    
    // #3 비밀번호 제외하고 반환
    const { password: _, ...result } = user;
    return result;
  }
}
